ovrlib/pa.py

```
@@ -0,0 +1,549 @@
+
+
+class InvalidAccessKeyError(Exception):
+    pass
+
+
+class ReadOnlyAccessKeyError(Exception):
+    pass
+
+
+class InvalidRegistrationError(Exception):
+    pass
+
+
+class InvalidDLError(Exception):
+    pass
+
+
+class InvalidSignatureError(Exception):
+    pass
+
+
+class Session:
+    def __init__(self, api_key, staging, language=0):
+        self.api_key = api_key
+        self.staging = staging
+        self.language = language
+
+    def get_url(self, action):
+        if self.staging:
+            url = STAGING_URL
+        else:
+            url = PROD_URL
+        url += f"?JSONv2&sysparm_AuthKey={self.api_key}&sysparm_action={action}&sysparm_Language={self.language}"
+        return url
+
+    def do_request(self, action, data=None):
+        url = self.get_url(action)
+        if data:
+            response = requests.post(
+                url,
+                headers={
+                    "Content-Type": "application/json",
+                    "Cache-Control": "no-cache",
+                },
+                data=data,
+            )
+        else:
+            response = requests.get(url)
+
+        logger.debug(f"{action} {response.status_code}")
+
+        if response.status_code != 200:
+            raise InvalidRegistrationError(f"HTTP status code {response.status_code}")
+
+        xml_str = json.loads(response.text)
+        root = etree.fromstring(xml_str)
+
+        # check for errors only
+        if root.tag == "RESPONSE":
```

We were seeing that incomplete requests would return something without an error, but also no registration ID, right? Should we throw an error for that case?